File: nodejs-mvc/layered-architecture/src/data/repositories/PostgresProductRepository.ts
import { Pool } from 'pg';
import { Product } from '../models/Product';

/**
 * @layer Data Layer
 * @description PostgreSQL implementation of the Product repository.
 * In Layered Architecture, this is a concrete class in the data layer
 * that handles persistence using PostgreSQL.
 *
 * It has the same public API as the in-memory ProductRepository,
 * allowing it to be used as a drop-in replacement.
 */
export class PostgresProductRepository {
  constructor(private readonly pool: Pool) {}

  async save(product: Product): Promise<Product> {
    const query = `
      INSERT INTO products (id, name, description, price, category, stock_quantity, sku, active, created_at, updated_at)
      VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
      RETURNING *
    `;
    const values = [
      product.id,
      product.name,
      product.description,
      product.price,
      product.category,
      product.stockQuantity,
      product.sku,
      product.active,
      product.createdAt,
      product.updatedAt,
    ];

    const result = await this.pool.query(query, values);
    return this.mapRowToProduct(result.rows[0]);
  }

  async findById(id: string): Promise<Product | null> {
    const query = 'SELECT * FROM products WHERE id = $1';
    const result = await this.pool.query(query, [id]);

    if (result.rows.length === 0) return null;
    return this.mapRowToProduct(result.rows[0]);
  }

  async findAll(page: number, size: number): Promise<{ products: Product[]; total: number }> {
    const countQuery = 'SELECT COUNT(*) FROM products WHERE active = true';
    const countResult = await this.pool.query(countQuery);
    const total = parseInt(countResult.rows[0].count);

    const offset = page * size;
    const query = 'SELECT * FROM products WHERE active = true ORDER BY created_at DESC LIMIT $1 OFFSET $2';
    const result = await this.pool.query(query, [size, offset]);

    const products = result.rows.map((row) => this.mapRowToProduct(row));
    return { products, total };
  }

  async findBySku(sku: string): Promise<Product | null> {
    const query = 'SELECT * FROM products WHERE sku = $1';
    const result = await this.pool.query(query, [sku]);

    if (result.rows.length === 0) return null;
    return this.mapRowToProduct(result.rows[0]);
  }

  async search(
    query?: string,
    category?: string,
    minPrice?: number,
    maxPrice?: number
  ): Promise<Product[]> {
    let sql = 'SELECT * FROM products WHERE active = true';
    const values: any[] = [];
    let paramIndex = 1;

    if (query) {
      sql += ` AND (LOWER(name) LIKE $${paramIndex} OR LOWER(description) LIKE $${paramIndex})`;
      values.push(`%${query.toLowerCase()}%`);
      paramIndex++;
    }

    if (category) {
      sql += ` AND LOWER(category) = $${paramIndex}`;
      values.push(category.toLowerCase());
      paramIndex++;
    }

    if (minPrice !== undefined) {
      sql += ` AND price >= $${paramIndex}`;
      values.push(minPrice);
      paramIndex++;
    }

    if (maxPrice !== undefined) {
      sql += ` AND price <= $${paramIndex}`;
      values.push(maxPrice);
      paramIndex++;
    }

    sql += ' ORDER BY created_at DESC';

    const result = await this.pool.query(sql, values);
    return result.rows.map((row) => this.mapRowToProduct(row));
  }

  async update(product: Product): Promise<Product> {
    const query = `
      UPDATE products
      SET name = $2, description = $3, price = $4, category = $5,
          stock_quantity = $6, sku = $7, active = $8, updated_at = $9
      WHERE id = $1
      RETURNING *
    `;
    const values = [
      product.id,
      product.name,
      product.description,
      product.price,
      product.category,
      product.stockQuantity,
      product.sku,
      product.active,
      product.updatedAt,
    ];

    const result = await this.pool.query(query, values);
    return this.mapRowToProduct(result.rows[0]);
  }

  async delete(id: string): Promise<void> {
    const query = 'DELETE FROM products WHERE id = $1';
    await this.pool.query(query, [id]);
  }

  private mapRowToProduct(row: any): Product {
    return {
      id: row.id,
      name: row.name,
      description: row.description,
      price: parseFloat(row.price),
      category: row.category,
      stockQuantity: row.stock_quantity,
      sku: row.sku,
      active: row.active,
      createdAt: new Date(row.created_at),
      updatedAt: new Date(row.updated_at),
    };
  }
}

File: python-mvc/layered-architecture/app/data/postgres_product_repository.py
"""
Data Layer: PostgreSQL Product Repository.

Concrete repository implementation using PostgreSQL via asyncpg.
Provides the same API as InMemoryProductRepository for interchangeability.

KEY DIFFERENCE FROM HEXAGONAL ARCHITECTURE:
- Hexagonal: Both InMemory and Postgres repositories implement the same
  abstract port (ProductRepositoryPort). Swapping is done via dependency injection.
- Layered: Both repositories are concrete classes with the same method signatures
  (duck typing). Swapping requires changing the wiring code in main.py.
  There is no formal contract (interface) guaranteeing API compatibility.

This is the trade-off of layered architecture: simpler code, but the "contract"
between layers is implicit rather than explicit.
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional

import asyncpg

from app.data.models import Product


class PostgresProductRepository:
    """
    PostgreSQL product repository.

    Uses asyncpg connection pool for efficient async database access.
    Same API as InMemoryProductRepository (duck typing, no shared interface).
    """

    def __init__(self, pool: asyncpg.Pool) -> None:
        self._pool = pool

    @staticmethod
    def _row_to_product(row: asyncpg.Record) -> Product:
        """Map a database row to a Product model."""
        return Product(
            id=str(row["id"]),
            name=row["name"],
            description=row["description"],
            price=float(row["price"]),
            category=row["category"],
            stock_quantity=row["stock_quantity"],
            sku=row["sku"],
            active=row["active"],
            created_at=row["created_at"],
            updated_at=row["updated_at"],
        )

    async def save(self, product: Product) -> Product:
        """Persist a new product to PostgreSQL."""
        query = """
            INSERT INTO products (id, name, description, price, category,
                                  stock_quantity, sku, active, created_at, updated_at)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
            RETURNING *
        """
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                query,
                product.id,
                product.name,
                product.description,
                product.price,
                product.category,
                product.stock_quantity,
                product.sku,
                product.active,
                product.created_at,
                product.updated_at,
            )
        return self._row_to_product(row)

    async def find_by_id(self, product_id: str) -> Optional[Product]:
        """Find a product by its unique identifier."""
        query = "SELECT * FROM products WHERE id = $1"
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(query, product_id)
        if row is None:
            return None
        return self._row_to_product(row)

    async def find_all(self, page: int, size: int) -> tuple[list[Product], int]:
        """
        Find all products with pagination.

        Returns a tuple of (products_list, total_count).
        """
        offset = (page - 1) * size

        count_query = "SELECT COUNT(*) FROM products"
        data_query = """
            SELECT * FROM products
            ORDER BY created_at DESC
            LIMIT $1 OFFSET $2
        """

        async with self._pool.acquire() as conn:
            total = await conn.fetchval(count_query)
            rows = await conn.fetch(data_query, size, offset)

        products = [self._row_to_product(row) for row in rows]
        return products, total

    async def find_by_sku(self, sku: str) -> Optional[Product]:
        """Find a product by its SKU."""
        query = "SELECT * FROM products WHERE sku = $1"
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(query, sku)
        if row is None:
            return None
        return self._row_to_product(row)

    async def search(
        self,
        query: Optional[str] = None,
        category: Optional[str] = None,
        min_price: Optional[float] = None,
        max_price: Optional[float] = None,
    ) -> list[Product]:
        """Search products by various criteria."""
        conditions: list[str] = []
        params: list = []
        param_idx = 1

        if query:
            conditions.append(
                f"(LOWER(name) LIKE ${param_idx} OR LOWER(description) LIKE ${param_idx})"
            )
            params.append(f"%{query.lower()}%")
            param_idx += 1

        if category:
            conditions.append(f"LOWER(category) = ${param_idx}")
            params.append(category.lower())
            param_idx += 1

        if min_price is not None:
            conditions.append(f"price >= ${param_idx}")
            params.append(min_price)
            param_idx += 1

        if max_price is not None:
            conditions.append(f"price <= ${param_idx}")
            params.append(max_price)
            param_idx += 1

        where_clause = " AND ".join(conditions) if conditions else "TRUE"
        sql = f"SELECT * FROM products WHERE {where_clause} ORDER BY name"

        async with self._pool.acquire() as conn:
            rows = await conn.fetch(sql, *params)

        return [self._row_to_product(row) for row in rows]

    async def update(self, product: Product) -> Product:
        """Update an existing product in PostgreSQL."""
        query = """
            UPDATE products
            SET name = $2, description = $3, price = $4, category = $5,
                stock_quantity = $6, sku = $7, active = $8, updated_at = $9
            WHERE id = $1
            RETURNING *
        """
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                query,
                product.id,
                product.name,
                product.description,
                product.price,
                product.category,
                product.stock_quantity,
                product.sku,
                product.active,
                product.updated_at or datetime.utcnow(),
            )
        return self._row_to_product(row)

    async def delete(self, product_id: str) -> None:
        """Delete a product by its identifier."""
        query = "DELETE FROM products WHERE id = $1"
        async with self._pool.acquire() as conn:
            await conn.execute(query, product_id)
